akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -313,6 +317,118 @@ private[stream] class VirtualPublisher[T]() extends Publisher[T] {
 /**
  * INTERNAL API
  */
+private[stream] case class MaterializedValueSource[M](
+  shape: SourceShape[M] = SourceShape[M](new Outlet[M]("Materialized.out")),
+  attributes: OperationAttributes = OperationAttributes.name("Materialized")) extends StreamLayout.Module {
+
+  override def subModules: Set[Module] = Set.empty
+  override def withAttributes(attr: OperationAttributes): Module = this.copy(shape = amendShape(attr), attributes = attr)
+  override def carbonCopy: Module = this.copy(shape = SourceShape(new Outlet[M]("Materialized.out")))
+
+  override def replaceShape(s: Shape): Module =
+    if (s == shape) this
+    else throw new UnsupportedOperationException("cannot replace the shape of MaterializedValueSource")
+
+  def amendShape(attr: OperationAttributes): SourceShape[M] = {
+    attr.nameOption match {
+      case None ⇒ shape
+      case s: Some[String] if s == attributes.nameOption ⇒ shape
+      case Some(name) ⇒ shape.copy(outlet = new Outlet(name + ".out"))
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] object MaterializedValuePublisher {
+  final val NotRequested = 0
+  final val Requested = 1
+  final val Completed = 2
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] class MaterializedValuePublisher extends Publisher[Any] {
+  import MaterializedValuePublisher._
+
+  private val value = new AtomicReference[AnyRef](null)
+  private val registeredSubscriber = new AtomicReference[Subscriber[_ >: Any]](null)
+  private val requestState = new AtomicInteger(NotRequested)
+
+  private def close(): Unit = {
+    requestState.set(Completed)
+    value.set(null)
+    registeredSubscriber.set(null)
+  }
+
+  private def tryOrClose(block: ⇒ Unit): Unit = {
+    try block catch {
+      case v: ReactiveStreamsCompliance.SpecViolation ⇒
+        close()
+      // What else can we do here?
+      case NonFatal(e) ⇒
+        val sub = registeredSubscriber.get()
+        if ((sub ne null) &&
+          requestState.compareAndSet(NotRequested, Completed) || requestState.compareAndSet(Requested, Completed)) {
+          sub.onError(e)
+        }
+        close()
+        throw e
+    }
+  }
+
+  def setValue(m: Any): Unit =
+    tryOrClose {
+      if (value.compareAndSet(null, m.asInstanceOf[AnyRef]) && requestState.get() == Requested)
```

Comment:
Is this safe in combination with the below

```scala
if (requestState.compareAndSet(NotRequested, Requested)) {
  val m = value.get()
  f (m ne null) pushAndClose(m)
}
```

I.e. when both run concurrently is it guaranteed that at least one of the `pushAndClose` terms fire? If its already safe, it may warrant an explaining comment. It seems missing both pushAndClose is prevented by allowing it to be called twice, instead, if both `compareAndSet` run concurrently and finish both before the respective other atomic accesses happen.